modal_proto/api.proto

```diff
@@ -1285,6 +1338,44 @@ message FunctionMapResponse {
   repeated FunctionPutInputsResponseItem pipelined_inputs = 2;
 }
 
+message FunctionMeta {
+  // Subset of fields from Function message above that will get deprecated there
+  // and made authoritative here. Represents metadata about the function itself,
+  // stored inline with the function object, capturing data not specific to any
+  // particular underlying task. Something specific would be the kind of GPU
+  // used, for instance.
+
+  string module_name = 1;
+  string function_name = 2;
+
+  Function.FunctionType function_type = 3;
+
+  // Scheduling related fields.
+  uint32 warm_pool_size = 4;
+  uint32 concurrency_limit = 5;
+  uint32 task_idle_timeout_secs = 6;
```

Comment:
Note that I'm not certain these fall cleanly into the "general"/"specific: dichotomy you define above and I suspect that users will be confused about what is and isn't possible to let vary.

Also, not sure how much you've thought about it, but seems like there could be some weird interactions between the scheduler parameters and the "gpu fallback" behavior. e.g. if you have a warm pool, what gpu types do we spin up for the pool? Do we take the first available — potentially leaving you with an idle but less-preferred type? Or do we try to keep your most preferred type warm? I might have similar questions about the maximum limit — if you are at your concurrency limit while using some less-preferred types when a more-preferred type would be available, are you just stuck there, or should try to migrate your usage to align with your preferences?

Reply:
For warm pools, etc. we'll generally scale up to take the first available but then consolidate it away once we have the most preferred type. Ditto for being stuck at the concurrency level - it would work the same way if you had a a blue task (from an old fn version) hogging up your limited concurrency slots - we'd still spin up more, to replace the blue, instead of having things get wedged there.